crates/iceberg/src/transform/temporal.rs

```diff
@@ -0,0 +1,669 @@
+            .unary(|v| 12 * (v - 1970));
+        let month_array = month_dyn(&input)
+            .map_err(|err| Error::new(crate::ErrorKind::ArrowCompute, format!("{err}")))?;
+        Ok(Arc::<Int32Array>::new(
+            binary(
+                month_array.as_any().downcast_ref::<Int32Array>().unwrap(),
+                year_array.as_any().downcast_ref::<Int32Array>().unwrap(),
+                // Compute month from 1970-01-01, so minus 1 here.
+                |a, b| a + b - 1,
+            )
+            .unwrap(),
+        ))
+    }
+}
+
+/// Extract a date or timestamp day, as days from 1970-01-01
+pub struct Day;
+
+impl TransformFunction for Day {
+    fn transform(&self, input: ArrayRef) -> Result<ArrayRef> {
+        let res: Int32Array = match input.data_type() {
+            DataType::Timestamp(unit, _) => match unit {
+                datatypes::TimeUnit::Second => input
+                    .as_any()
+                    .downcast_ref::<TimestampSecondArray>()
+                    .unwrap()
+                    .unary(|v| -> i32 { (v as f64 * DAY_PER_SECOND) as i32 }),
+                datatypes::TimeUnit::Millisecond => input
+                    .as_any()
+                    .downcast_ref::<TimestampMillisecondArray>()
+                    .unwrap()
+                    .unary(|v| -> i32 { (v as f64 / 1000.0 * DAY_PER_SECOND) as i32 }),
+                datatypes::TimeUnit::Microsecond => input
+                    .as_any()
+                    .downcast_ref::<TimestampMicrosecondArray>()
+                    .unwrap()
+                    .unary(|v| -> i32 { (v as f64 / 1000.0 / 1000.0 * DAY_PER_SECOND) as i32 }),
+                datatypes::TimeUnit::Nanosecond => input
+                    .as_any()
+                    .downcast_ref::<TimestampNanosecondArray>()
+                    .unwrap()
+                    .unary(|v| -> i32 {
+                        (v as f64 / 1000.0 / 1000.0 / 1000.0 * DAY_PER_SECOND) as i32
+                    }),
+            },
+            DataType::Date32 => {
+                input
+                    .as_any()
+                    .downcast_ref::<Date32Array>()
+                    .unwrap()
+                    .unary(|v| -> i32 {
+                        datatypes::Date32Type::to_naive_date(v).num_days_from_ce()
+                            - EPOCH_DAY_FROM_CE
+                    })
+            }
+            DataType::Date64 => {
+                input
+                    .as_any()
+                    .downcast_ref::<Date64Array>()
+                    .unwrap()
+                    .unary(|v| -> i32 {
+                        datatypes::Date64Type::to_naive_date(v).num_days_from_ce()
+                            - EPOCH_DAY_FROM_CE
+                    })
+            }
+            _ => unreachable!(
+                "Should not call transform in Day with type {:?}",
+                input.data_type()
+            ),
+        };
+        Ok(Arc::new(res))
```

Comment:
Missing timestamp with timezone?

Reply:
> If a Timestamp column has a non-empty timezone value, its epoch is 1970-01-01 00:00:00 (January 1st 1970, midnight) in the UTC timezone (the Unix epoch), regardless of the Timestamp’s own timezone.

The timezone will not affect the epoch so we can use it directly. The result is always regard as in the UTC timezone.

https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html#timestamps-with-a-non-empty-timezone